src/components/Plans/PlanCalculator/reducers/plansByTypeReducer.test.js

```diff
@@ -72,53 +72,232 @@ describe('plansByTypeReducer', () => {
-    };
+  describe(`${PLANS_BY_TYPE_ACTIONS.SELECT_PLAN} action`, () => {
+    it('Should select a new plan and set previous selected discount when its index exists in new plan discounts ', () => {
+      // Arrange
 
-    // Act
-    const newState = plansByTypeReducer(initialState, action);
+      const discountA = {
+        id: 1,
+        billingCycle: 'monthly',
+        discountPercentage: 10,
+      };
 
-    // Assert
-    expect(newState).toEqual(
-      expect.objectContaining({
-        selectedPlanIndex,
-        selectedPlan: desiredPlan,
-      }),
-    );
+      const discountB = {
+        id: 2,
+        billingCycle: 'quarterly',
+        discountPercentage: 20,
+      };
+      const discountC = {
+        id: 3,
+        billingCycle: 'yearly',
+        discountPercentage: 30,
+      };
+
+      const discounts = [discountA, discountB, discountC];
+      const desiredPlan = { desiredPlan: true, billingCycleDetails: discounts };
+      const initialState = {
+        ...INITIAL_STATE_PLANS_BY_TYPE,
+        plansByType: [{}, {}, {}, {}, {}, desiredPlan, {}],
+        selectedDiscountIndex: 2,
+        discounts,
+      };
+      const selectedPlanIndex = 5;
+      const expectedSelectedDiscountIndex = 2;
+      const expectedDiscounts = [
+        {
+          id: 1,
+          subscriptionType: 'monthly',
+          numberMonths: 1,
+          discountPercentage: 10,
+        },
+        {
+          id: 2,
+          subscriptionType: 'quarterly',
+          numberMonths: 3,
+          discountPercentage: 20,
+        },
+        {
+          id: 3,
+          subscriptionType: 'yearly',
+          numberMonths: 12,
+          discountPercentage: 30,
+        },
+      ];
+
+      const action = {
+        type: PLANS_BY_TYPE_ACTIONS.SELECT_PLAN,
+        payload: selectedPlanIndex,
+      };
+
+      // Act
+      const newState = plansByTypeReducer(initialState, action);
+
+      // Assert
+      expect(newState).toEqual(
+        expect.objectContaining({
+          selectedPlanIndex,
+          selectedPlan: desiredPlan,
+          discounts: expectedDiscounts,
+          selectedDiscount: expectedDiscounts[2],
+          selectedDiscountIndex: expectedSelectedDiscountIndex,
+        }),
+      );
+    });
+
+    it('Should select a new plan and set first discount index when previous discount index does not exist in new plan discounts ', () => {
+      // Arrange
+
+      const discountA = {
+        id: 1,
+        billingCycle: 'monthly',
+        discountPercentage: 10,
+      };
+
+      const discountB = {
+        id: 2,
+        billingCycle: 'quarterly',
+        discountPercentage: 20,
+      };
+
+      const discounts = [discountA, discountB];
+      const desiredPlan = { desiredPlan: true, billingCycleDetails: discounts };
+      const initialState = {
+        ...INITIAL_STATE_PLANS_BY_TYPE,
+        plansByType: [{}, {}, {}, {}, {}, desiredPlan, {}],
+        selectedDiscountIndex: 2,
+        discounts,
```

Comment:
Este escenario no es realista porque `selectedDiscountIndex` apunta al tercer descuento mientras que `discounts` tiene solo dos.

Además falta el `selectedDiscount`.

Reply:
this case could happen, we talked about that. This case shows a scenary where selectedDiscountIndex in previous plan was 2. When user changes the plan, in the new one we only have an array with two positions. Position 2 does not exist. So what the reducer do it's set selectedDiscountIndex to 0 and selectedDiscount to first position of array.

Reply:
State selectedDiscount doesn't care because it's a discount of a previous plan. What we care it's selectedDiscountIndex.

Reply:
> this case could happen

No leí todo, pero no, no es un caso que pueda darse, ya que este es el estado anterior a la acción.